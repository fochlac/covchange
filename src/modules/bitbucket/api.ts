import * as request from 'request-promise-native'

import error from '../../utils/error'

const bitbucketurl = `${global.bitbucket}/rest/api/1.0/`
const Authorization = global.bitbucketKey
const { internalError } = error('Bitbucket API call failed: ')

export const Bitbucket = {
	get: (url: string): Promise<any> => apiCall({ url }),
	put: (url: string, body: any): Promise<any> => apiCall({ method: 'put', url, body }),
	post: (url: string, body: any): Promise<any> => apiCall({ method: 'post', url, body }),
}

interface ApiConfig {
	uri: string
	method: string
	json: boolean
	headers: { [key: string]: string }
	body?: any
}
interface ApiCallParameter {
	url: string
	method?: string
	body?: any
}
function apiCall({ url, method, body }: ApiCallParameter): Promise<any> {
	const config: ApiConfig = {
		uri: bitbucketurl + url,
		method,
		json: true,
		headers: {
			Authorization,
		},
	}

	if (body) {
		config.body = body
	}

	return request(config).catch(internalError(2, 'Error posting comment.'))
}
