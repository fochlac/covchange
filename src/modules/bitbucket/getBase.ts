import * as request from 'request-promise-native'

import error from '../../utils/error'

const { internalError } = error('bitbucket comment:')

const Authorization = global.bitbucketKey
const baseUrl = ({ repository: { slug, name }, pullRequestId }: Core.PullRequestRest) =>
	`${global.bitbucket}/rest/api/1.0/${slug}/repos/${name}/pull-requests/${pullRequestId}/`


export async function getBaseBranchFromPullrequest(pullRequest:Core.PullRequestRest):Promise<Core.BaseBranch> {
	const uri = baseUrl(pullRequest)
	try {
		const { toRef } = await request({
			uri,
			method: 'GET',
			json: true,
			headers: {
				Authorization,
			}
		})

		const slug = toRef.repository.project.type === 'NORMAL' ? 'projects' : 'users'

		return {
			name: toRef.displayId,
			repository: {
				name: toRef.repository.slug,
				slug: `${slug}/${toRef.repository.project.key}`
			}
		}

	}
	catch(err) {
		internalError(2, 'Error fetching comment.')(err)
	}
}
