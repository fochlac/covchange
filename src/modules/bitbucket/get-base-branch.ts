import { Bitbucket } from './bitbucket'

const baseUrl = ({ repository: { type, repo, project }, name }: Core.PullRequestRest) =>
	`${type}/${project}/repos/${repo}/pull-requests/${name}/`

export async function getBaseBranchFromPullrequest(pullRequest: Core.PullRequestRest): Promise<Core.BaseBranch> {
	const { toRef } = await Bitbucket.get(baseUrl(pullRequest))

	const type = toRef.repository.project.type === 'NORMAL' ? 'projects' : 'users'

	return {
		name: toRef.displayId,
		repository: {
			repo: toRef.repository.slug,
			type,
			project: toRef.repository.project.key,
		},
	}
}
