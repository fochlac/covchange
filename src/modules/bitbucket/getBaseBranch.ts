import * as request from 'request-promise-native'

import { Bitbucket } from './bitbucket'
import error from '../../utils/error'

const { internalError } = error('bitbucket comment:')

const baseUrl = ({ repository: { type, repo, project }, name }: Core.PullRequestRest) =>
	`${type}/${repo}/repos/${project}/pull-requests/${name}/`

export async function getBaseBranchFromPullrequest(pullRequest: Core.PullRequestRest): Promise<Core.BaseBranch> {
	try {
		const { toRef } = await Bitbucket.get(baseUrl(pullRequest))

		const type = toRef.repository.project.type === 'NORMAL' ? 'projects' : 'users'

		return {
			name: toRef.displayId,
			repository: {
				repo: toRef.repository.slug,
				type,
				project: toRef.repository.project.key,
			},
		}
	} catch (err) {
		internalError(2, 'Error fetching comment.')(err)
	}
}
