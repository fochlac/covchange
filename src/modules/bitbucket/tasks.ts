import { Bitbucket } from './bitbucket'

function createApiSlug(taskId?: Core.Id) {
	return `tasks/${taskId || ''}`
}

function updateTask(comment: Core.Comment, open?: boolean): Promise<Core.Id> {
	return Bitbucket.put(createApiSlug(comment.taskId), {
		id: comment.taskId,
		state: open ? 'OPEN' : 'RESOLVED',
	}).then(({ id }: { id: number }) => id.toString())
}

function createTask(comment: Core.Comment): Promise<Core.Id> {
	return Bitbucket.post(createApiSlug(), {
		text: 'Please add more unit tests.',
		anchor: {
			id: comment.commentId,
			type: 'COMMENT',
		},
	}).then(({ id }: { id: number }) => id.toString())
}

export function openTask(comment: Core.Comment): Promise<Core.Id> {
	return comment.taskId ? updateTask(comment, true) : createTask(comment)
}

export function resolveTask(comment: Core.Comment) {
	return comment.taskId ? updateTask(comment, false) : Promise.resolve(comment.taskId)
}
