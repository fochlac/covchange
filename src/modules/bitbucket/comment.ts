import { openTask, resolveTask } from './tasks'

import { Bitbucket } from './bitbucket'
import { commentDb } from '../db/comments'
import { createCommentObject } from '../../utils/parsing/create-comment'
import { diffReports } from '../../utils/parsing/diff'

function createApiSlug({ repository: { type, repo, project }, name }: Core.PullRequest, commentId?: Core.Id) {
	return `${type}/${project}/repos/${repo}/pull-requests/${name}/comments/${commentId || ''}`
}

export async function writeReportToBitbucket(base: Core.Branch, pr: Core.PullRequest) {
	const diff = diffReports(base.reports[0], pr.reports[0])
	const existingComment = await commentDb.get(pr.repository, pr.name)
	const comment = createCommentObject(diff, existingComment, pr.lcov)

	const { id, version }: { id: number; version: number } = existingComment
		? await Bitbucket.get(createApiSlug(pr, existingComment.commentId))
			.then(({ version, id }) => Bitbucket.put(createApiSlug(pr, id), { ...comment, version }))
		: await Bitbucket.post(createApiSlug(pr), comment)

	const newComment: Core.Comment = {
		...existingComment,
		commentId: id.toString(),
		version,
	}

	if (pr.task && diff.total.diff.statementCov < 0) {
		newComment.taskId = await openTask(newComment)
	} else if (newComment.taskId && diff.total.diff.statementCov >= 0) {
		await resolveTask(newComment)
	}

	return commentDb.set(pr.repository, pr.name, newComment)
}
