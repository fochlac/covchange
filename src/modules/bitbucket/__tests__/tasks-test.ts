import { openTask, resolveTask } from '../tasks';

import { Bitbucket } from '../bitbucket'

jest.mock('../../../utils/logger')
jest.mock('../bitbucket', () => ({
	Bitbucket: {
		put: jest.fn(() => Promise.resolve({ id: 42, version: 1 })),
		post: jest.fn(() => Promise.resolve({ id: 23, version: 1 })),
	},
}))

const testcomment = { commentId: 'commentId' }
const testcommentWithTask = { commentId: 'commentId', taskId: '12' }

describe('task', () => {

	afterEach(() => {
		// @ts-ignore
		Bitbucket.put.mockClear()
		// @ts-ignore
		Bitbucket.post.mockClear()
	})

	it('should create a task', async () => {
		expect(await openTask(testcomment)).toEqual('23')
		expect(Bitbucket.put).toBeCalledTimes(0)
		expect(Bitbucket.post).toBeCalledTimes(1)
		// @ts-ignore
		expect(Bitbucket.post.mock.calls).toMatchSnapshot()
	})

	it('should update existing task to open', async () => {
		expect(await openTask(testcommentWithTask)).toEqual('42')
		expect(Bitbucket.put).toBeCalledTimes(1)
		expect(Bitbucket.post).toBeCalledTimes(0)
		// @ts-ignore
		expect(Bitbucket.put.mock.calls).toMatchSnapshot()
	})

	it('should update existing task to resolved', async () => {
		expect(await resolveTask(testcommentWithTask)).toEqual('42')
		expect(Bitbucket.put).toBeCalledTimes(1)
		expect(Bitbucket.post).toBeCalledTimes(0)
		// @ts-ignore
		expect(Bitbucket.put.mock.calls).toMatchSnapshot()
	})

	it('should do nothing', async () => {
		expect(await resolveTask(testcomment)).toBeUndefined()
		expect(Bitbucket.put).toBeCalledTimes(0)
		expect(Bitbucket.post).toBeCalledTimes(0)
	})
})
