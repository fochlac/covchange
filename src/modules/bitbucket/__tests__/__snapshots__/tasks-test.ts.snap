// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task should create a task 1`] = `
Array [
  Array [
    "tasks/",
    Object {
      "anchor": Object {
        "id": "commentId",
        "type": "COMMENT",
      },
      "text": "Please add more unit tests.",
    },
  ],
]
`;

exports[`task should update existing task to open 1`] = `
Array [
  Array [
    "tasks/12",
    Object {
      "id": "12",
      "state": "OPEN",
    },
  ],
]
`;

exports[`task should update existing task to resolved 1`] = `
Array [
  Array [
    "tasks/12",
    Object {
      "id": "12",
      "state": "RESOLVED",
    },
  ],
]
`;
